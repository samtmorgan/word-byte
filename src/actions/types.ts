export interface Result {
  created: number;
  pass: boolean;
}

export enum WordOwner {
  PLATFORM = 'platform',
  USER = 'user',
}
export interface Word {
  word: string;
  wordId: string;
  owner: WordOwner;
  results: Result[];
}

export interface WordSet {
  wordSetId: string;
  createdAt: number;
  wordIds: string[];
}

export interface DbUser {
  _id: string;
  userAuthId: string;
  userPlatformId: string;
  createdAt: number;
  wordSets: WordSet[];
  words: Word[];
}

export type AuthUser = {
  userAuthId: string;
  username: string;
};

export interface User extends DbUser {
  username: string;
}

export type NewIdDbUser = Omit<DbUser, '_id'>;
