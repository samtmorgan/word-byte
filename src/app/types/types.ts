export type ButtonType = {
  onClick: () => void;
  label: string;
  disabled?: boolean;
  type?: 'button' | 'submit';
};
