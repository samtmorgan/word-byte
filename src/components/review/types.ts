import React from 'react';
import { Word } from '../../actions/getUser';

type ReviewProps = {
  currentWords: Word[];
};

type LocalResults = {
  pass: boolean | null;
  wordId: string;
  word: string;
}[];

enum ReviewLifecycle {
  IN_PROGRESS = 'inProgress',
  COMPLETE = 'complete',
}

type ReviewButtonsProps = {
  reviewLifecycle: ReviewLifecycle;
  setReviewLifecycle: React.Dispatch<React.SetStateAction<ReviewLifecycle>>;
};

export { ReviewLifecycle };
export type { ReviewProps, LocalResults, ReviewButtonsProps };
