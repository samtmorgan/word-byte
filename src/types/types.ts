// export type UserType = {
//   name: string;
//   uuid: string;
//   words: UserWordType[];
// } | null;

// export type UserType = {
//   name: string;
//   uuid: string;
//   words: {
//     current: string[] | null;
//     history: object[] | null;
//     customWords: UserWordType[];
//   };
// } | null;

export type UserType = {
  name: string;
  uuid: string;
  words: {
    wordSets: string[][];
    attempts: object[];
    customWords: UserWordType[];
  };
} | null;

export type MockUserType = {
  name: string;
  uuid: string;
  words: {
    wordSets: string[][];
    attempts: object[];
    customWords: UserWordType[];
  };
};

export type TestLifecycleType = 'notStarted' | 'test' | 'review' | 'revise' | 'finished' | 'cancelled';

export type ContextType = {
  loading: boolean;
  setLoading: (loading: boolean) => void;
  error: boolean;
  setError: (error: boolean) => void;
  user: UserType;
  setUser: (user: UserType) => void;
  testWords: string[];
  //   sessionWords: SessionWordType[] | null;
  //   setSessionWords: (sessionWords: SessionWordType[] | null) => void;
  //   sessionWords: string[] | null;
  //   setSessionWords: (sessionWords: string[] | null) => void;
  //   testLifecycle: string | null;
  //   setTestLifecycle: (testLifecycle: TestLifecycleType | null) => void;
};

export type UserWordType = {
  word: string;
  uuid: string;
  //   owner: 'platform' | 'user';
  owner: string;

  //   current: boolean;
  //   attempts:
  //     | {
  //         timestamp: number;
  //         pass: boolean;
  //       }[]
  //     | null;
};

export interface SessionWordType extends UserWordType {
  correct: boolean;
}

export type InputType = {
  value: string;
  onChange: (value: string) => void;
  name: string;
  placeholder?: string;
  label?: string;
};

export type ResultType = {
  word: string;
  correct: boolean;
};
