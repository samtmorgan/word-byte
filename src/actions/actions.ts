import client from '../lib/mongoClient';

interface Result {
  created: number;
  pass: boolean;
}

interface Word {
  word: string;
  wordId: string;
  owner: 'platform' | 'user';
  results: Result[];
}

interface WordSet {
  wordSetId: string;
  created: number;
  wordIds: string[];
}

export interface User {
  _id: string;
  userAuthId: string;
  userPlatformId: string;
  userName: string;
  createdAt: 1735938406366;
  wordSets: WordSet[];
  words: Word[];
}

export async function getUser(userAuthId: string): Promise<User | null | 'error'> {
  try {
    const mongoClient = await client.connect();
    const db = mongoClient.db('wordByteTest');
    const user = (await db.collection('users').findOne({ userAuthId })) as User | null;
    return user;
  } catch (e) {
    console.error(e);
    return 'error';
  }
}
