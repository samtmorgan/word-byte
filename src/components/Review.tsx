'use client';

import React, { useCallback, useEffect, useState } from 'react';
import { useAppContext } from '../context/AppContext';
import Fireworks from './Fireworks';
import { ResultType } from '../types/types';

export default function Review() {
  const { sessionWords, loading } = useAppContext();
  const [isVisible, setIsVisible] = useState<boolean>(false);
  const [results, setResults] = useState<ResultType[]>(sessionWords?.map(word => ({ word, correct: false })) || []);

  const handleWordClick = useCallback(
    (index: number) => {
      if (!sessionWords) return;
      const newResults = [...results];
      newResults[index].correct = !newResults[index].correct;
      setResults(newResults);
    },
    [results, sessionWords],
  );
  useEffect(() => {
    if (results?.every(({ correct }) => correct)) {
      setIsVisible(true);
      setTimeout(() => setIsVisible(false), 10000);
    }
  }, [results]);

  if (loading) return <div>Loading...</div>;

  if (!sessionWords) return <div>No words</div>;

  return (
    <>
      <div className="page-container">
        <section className="review">
          <h2>Click the words you got right ✓</h2>
          {results.map((wordData, index) => (
            <button
              className={`button cool-border-with-shadow ${wordData.correct && 'correct-word'}`}
              type="button"
              onClick={() => handleWordClick(index)}
              key={wordData.word}
            >
              {`${wordData.word} ${wordData.correct ? '✓' : ''}`}
            </button>
          ))}
        </section>
      </div>
      {/* {isVisible && <Confetti />} */}
      {isVisible && <Fireworks />}
    </>
  );
}
