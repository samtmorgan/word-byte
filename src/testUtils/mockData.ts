import { DbUser } from '../actions/getUser';
import { MockUserType } from '../types/types';

// ~~~~ current below \/

export const mockAuthUserId = 'auth123';
export const mockUsername = 'testUser';
export const mockDbUser: DbUser = {
  _id: '1',
  userAuthId: mockAuthUserId,
  userPlatformId: 'platform123',
  createdAt: 1735938406366,
  wordSets: [
    {
      wordSetId: 'mockWordSetId',
      createdAt: 123,
      wordIds: ['mockWordId'],
    },
  ],
  words: [{ word: 'mockWord', wordId: 'mockWordId', owner: 'platform', results: [] }],
};

export const mockUser = {
  ...mockDbUser,
  username: mockUsername,
};

export const mockCurrentWords = [
  {
    word: 'accident',
    wordId: 'd659c3a4-4ea2-4619-b4da-53b6550d925f',
    owner: 'platform',
    results: [],
  },
  {
    word: 'accidentally',
    wordId: 'b02b5c18-1ed6-40e5-bf4b-dc268a1b8866',
    owner: 'platform',
    results: [],
  },
  {
    word: 'actual',
    wordId: 'e13e18b1-7c04-4dae-8a02-6898848d168b',
    owner: 'platform',
    results: [],
  },
  {
    word: 'actually',
    wordId: '1ed7098e-e2a7-4607-940c-cdb7f49b3bc0',
    owner: 'platform',
    results: [],
  },
  {
    word: 'address',
    wordId: '8ec8b1ff-50a1-4152-bae2-01ff3e5450ad',
    owner: 'platform',
    results: [],
  },
  {
    word: 'answer',
    wordId: 'fbc6c6b1-df0b-4015-94c9-01207ceb4471',
    owner: 'platform',
    results: [],
  },
  {
    word: 'believe',
    wordId: '26dd4d7a-045c-4103-8461-1466bd4b84c1',
    owner: 'platform',
    results: [],
  },
  {
    word: 'bicycle',
    wordId: 'a86b4f9e-7027-4940-9345-fed0951c87b0',
    owner: 'platform',
    results: [],
  },
  {
    word: 'believe',
    wordId: '26dd4d7a-045c-4103-8461-1466bd4b84c1',
    owner: 'platform',
    results: [],
  },
  {
    word: 'bicycle',
    wordId: 'a86b4f9e-7027-4940-9345-fed0951c87b0',
    owner: 'platform',
    results: [],
  },
];

// ~;~~~ current above /\

export const wordSetText = [
  'illogical',
  'illiterate',
  'illegal',
  'illegible',
  'impatient',
  'impractical',
  'disqualify',
  'disobey',
  'incorrect',
  'incomplete',
];

export const mockUserWords = [
  {
    word: 'illogical',
    uuid: '1d4ce9f5-4a58-48ea-a7fe-4a97dda52601',
    owner: 'platform',
  },
  {
    word: 'illiterate',
    uuid: '1c94337d-be0d-409e-a422-5f1576bb9764',
    owner: 'platform',
  },
  {
    word: 'illegal',
    uuid: '2c35689c-bbea-45ff-8245-50d6f9e4f105',
    owner: 'platform',
  },
  {
    word: 'illegible',
    uuid: 'cd7514c8-f2e3-4c8d-a8d7-d5584ce3ee97',
    owner: 'platform',
  },
  {
    word: 'impatient',
    uuid: '583770e4-1356-4639-a17a-342ad43daad9',
    owner: 'platform',
  },
  {
    word: 'impractical',
    uuid: 'e768960c-77ab-4212-a8d7-ef426b1e6201',
    owner: 'platform',
  },
  {
    word: 'disqualify',
    uuid: '065eeef7-b439-4450-8669-d5afc0daf8de',
    owner: 'platform',
  },
  {
    word: 'disobey',
    uuid: '1f6d9fec-f515-448c-b858-564459808183',
    owner: 'platform',
  },
  {
    word: 'incorrect',
    uuid: '3c291577-2883-4bb8-abc9-1ced57cc6dbc',
    owner: 'platform',
  },
  {
    word: 'incomplete',
    uuid: '8db538e0-a490-4fc9-b45f-b809a19668fc',
    owner: 'platform',
  },
];

export const oldMockUser: MockUserType = {
  name: 'Ginny',
  uuid: 'b8765153-9c08-4301-869c-d810453408ca',
  words: {
    wordSets: [wordSetText],
    attempts: [],
    customWords: mockUserWords,
  },
};

export const mockTestWords = [
  'recommend',
  'relevant',
  'restaurant',
  'rhyme',
  'rhythm',
  'sacrifice',
  'secretary',
  'shoulder',
  'signature',
  'sincere',
];
