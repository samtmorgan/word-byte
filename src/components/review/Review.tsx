'use client';

import React, { useState, JSX } from 'react';
import { Word } from '../../actions/getUser';
import { Complete } from './components/Complete';
import { InProgress } from './components/InProgress';
import { LocalResults, ReviewButtonsProps, ReviewLifecycle, ReviewProps } from './types';

function buildLocalResults(currentWords: Word[]): LocalResults {
  return currentWords.map(word => ({ ...word, pass: null }));
}

const ReviewButtons = ({ reviewLifecycle, setReviewLifecycle }: ReviewButtonsProps): JSX.Element => (
  <div className="button-container">
    {reviewLifecycle === ReviewLifecycle.IN_PROGRESS ? (
      <button
        className="button cool-border-with-shadow finish-button"
        type="button"
        onClick={() => setReviewLifecycle(ReviewLifecycle.COMPLETE)}
      >
        üèÅ Finish
      </button>
    ) : (
      <a className="button cool-border-with-shadow finish-button" href="/">
        üè† Go to home
      </a>
    )}
  </div>
);

export default function Review({ currentWords }: ReviewProps) {
  const [results, setResults] = useState<LocalResults>(buildLocalResults(currentWords || []));
  const [reviewLifecycle, setReviewLifecycle] = useState<ReviewLifecycle>(ReviewLifecycle.IN_PROGRESS);

  if (!results || results.length === 0) return <div>üôÅ No words here yet</div>;

  return (
    <div className="page-container">
      <section className="review">
        {reviewLifecycle === ReviewLifecycle.IN_PROGRESS ? (
          <InProgress results={results} setResults={setResults} />
        ) : (
          <Complete results={results} />
        )}
        <ReviewButtons reviewLifecycle={reviewLifecycle} setReviewLifecycle={setReviewLifecycle} />
      </section>
    </div>
  );
}
