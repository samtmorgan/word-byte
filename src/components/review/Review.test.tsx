import React from 'react';
import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';
import { render, screen } from '@testing-library/react';
import Review from './Review';
import { mockCurrentWords } from '../../testUtils/mockData';
import { Word } from '../../actions/types';

jest.mock('./components/InProgress', () => ({
  InProgress: () => <div>MockInProgress</div>,
}));
jest.mock('./components/Complete', () => ({
  Complete: () => <div>MockComplete</div>,
}));

describe.only('Review component', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should render message when there are no words in the set', () => {
    render(<Review currentWords={[]} />);

    expect(screen.getByText(/🙁 No words here yet/)).toBeInTheDocument();
  });

  it('should render the in progress view on mount', async () => {
    render(<Review currentWords={mockCurrentWords as Word[]} />);

    expect(screen.getByText(/MockInProgress/)).toBeInTheDocument();
  });

  it('should render the review complete view when the user clicks the finish button', async () => {
    render(<Review currentWords={mockCurrentWords as Word[]} />);
    const user = userEvent.setup();
    const button = screen.getByRole('button', { name: '🏁 Finish' });
    await user.click(button);
    expect(screen.getByText(/MockComplete/)).toBeInTheDocument();
    expect(screen.queryByText(/MockInProgress/)).not.toBeInTheDocument();
  });

  it.skip('should render expected content when there are words in the set', () => {
    const { container } = render(<Review currentWords={mockCurrentWords as Word[]} />);

    expect(screen.getByText(/Click the words you got right ✓/)).toBeInTheDocument();
    expect(screen.getByRole('list')).toBeInTheDocument();
    expect(screen.getAllByRole('listitem').length).toEqual(mockCurrentWords.length);
    expect(container).toMatchSnapshot();
  });

  it.skip('should respond to user interaction', async () => {
    render(<Review currentWords={mockCurrentWords.slice(0, 1) as Word[]} />);

    const firstWord = mockCurrentWords[0].word;
    const user = userEvent.setup();
    let button = screen.getAllByRole('button')[0];

    await user.click(button);

    expect(screen.getByText(`${firstWord} ✓`)).toBeInTheDocument();

    await user.click(button);

    [button] = screen.getAllByRole('button');

    expect(screen.queryByText(`${firstWord} `)).toBeInTheDocument();
    expect(screen.queryByText(firstWord)).toBeInTheDocument();
  });

  it.skip('should render the fireworks when the user has reviewed all the words and they are correct', async () => {
    render(<Review currentWords={mockCurrentWords.slice(0, 1) as Word[]} />);
    const button = screen.getAllByRole('button')[0];
    const user = userEvent.setup();

    await user.click(button);

    expect(screen.getByText(/Confetti/)).toBeInTheDocument();
  });
});
