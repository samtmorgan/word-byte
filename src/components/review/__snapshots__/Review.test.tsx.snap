// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Review component should render expected content when there are words in the set 1`] = `
<div>
  <div
    class="pageContainer"
  >
    <section
      class="review"
    >
      <h1>
        Time to check your spellings!
      </h1>
      <p>
        Click the words you got right ‚úì
      </p>
      <ol>
        <li
          class="word-list"
        >
          <button
            class=""
            type="button"
          >
            accident 
          </button>
        </li>
        <li
          class="word-list"
        >
          <button
            class=""
            type="button"
          >
            accidentally 
          </button>
        </li>
        <li
          class="word-list"
        >
          <button
            class=""
            type="button"
          >
            actual 
          </button>
        </li>
        <li
          class="word-list"
        >
          <button
            class=""
            type="button"
          >
            actually 
          </button>
        </li>
        <li
          class="word-list"
        >
          <button
            class=""
            type="button"
          >
            address 
          </button>
        </li>
        <li
          class="word-list"
        >
          <button
            class=""
            type="button"
          >
            answer 
          </button>
        </li>
        <li
          class="word-list"
        >
          <button
            class=""
            type="button"
          >
            believe 
          </button>
        </li>
        <li
          class="word-list"
        >
          <button
            class=""
            type="button"
          >
            bicycle 
          </button>
        </li>
        <li
          class="word-list"
        >
          <button
            class=""
            type="button"
          >
            breathe 
          </button>
        </li>
        <li
          class="word-list"
        >
          <button
            class=""
            type="button"
          >
            build 
          </button>
        </li>
      </ol>
      <div
        class="button-container"
      >
        <button
          class="finish-button"
          type="button"
        >
          üèÅ Finish
        </button>
      </div>
    </section>
  </div>
</div>
`;
